### ISSUE TITLE
command line usage: option -q is not mentioned

### ISSUE THREAD

**Issue by yecril71pl** [rpm-software-management/rpm#1473]
{ rpm; }

> Składnia: rpm [-afgplsiv?] [-a|--all] [-f|--file] [-g|--group] [-p|--package]
>        [--pkgid] [--hdrid] [--triggeredby] [--whatconflicts] [--whatrequires]
>        [--whatobsoletes] [--whatprovides] [--whatrecommends] [--whatsuggests]
>        [--whatsupplements] [--whatenhances] [--nomanifest] [-c|--configfiles]
>        [-d|--docfiles] [-L|--licensefiles] [-A|--artifactfiles] [--noghost]
>        [--noconfig] [--noartifact] [--dump] [-l|--list]
>        [--queryformat=QUERYFORMAT] [-s|--state] [--nofiledigest] [--nofiles]
>        [--nodeps] [--noscript] [--allfiles] [--allmatches] [--badreloc]
>        [-e|--erase=<pakiet>+] [--excludedocs] [--excludepath=<ścieżka>] [--force]
>        [-F|--freshen=<plik-pakietu>+] [-h|--hash] [--ignorearch] [--ignoreos]
>        [--ignoresize] [--noverify] [-i|--install] [--justdb] [--nodeps]
>        [--nofiledigest] [--nocontexts] [--nocaps] [--noorder] [--noscripts]
>        [--notriggers] [--oldpackage] [--percent] [--prefix=<katalog>]
>        [--relocate=<poprzednia>=<nowa>] [--replacefiles] [--replacepkgs] [--test]
>        [-U|--upgrade=<plik-pakietu>+] [--reinstall=<plik-pakietu>+]
>        [-D|--define="MAKRO WYRAŻENIE"] [--undefine=MAKRO] [-E|--eval="WYRAŻENIE"]
>        [--target=PROCESOR-PRODUCENT-SYSTEM_OPERACYJNY] [--macros=<PLIK:…>]
>        [--load=<PLIK>] [--noplugins] [--nodigest] [--nosignature]
>        [--rcfile=<PLIK:…>] [-r|--root=ROOT] [--dbpath=KATALOG] [--querytags]
>        [--showrc] [--quiet] [-v|--verbose] [--version] [-?|--help] [--usage]
>        [--scripts] [--setperms] [--setugids] [--setcaps] [--restore]
>        [--conflicts] [--obsoletes] [--provides] [--requires] [--recommends]
>        [--suggests] [--supplements] [--enhances] [--info] [--changelog]
>        [--changes] [--xml] [--triggers] [--filetriggers] [--last] [--dupes]
>        [--filesbypkg] [--fileclass] [--filecolor] [--fileprovide] [--filerequire]
>        [--filecaps]

**Comment by xsuchy** [comment]
How you get this output? I see `-q` in both `man rpm` and `rpm -h` - but output of both differs from what you posted.

**Comment by xsuchy** [comment]
Ahh. It is simple output of `rpm` without any param. And yes, it is not there.

**Comment by dmnks** [comment]
This is a bit more complicated than one would think. The `--query` option is deliberately hidden from the usage/help output because there simply isn't a good help section to house it:

```
$ rpm --help
[...]
Query/Verify package selection options:
Query/Verify file selection options:
Query options (with -q or --query):
Verify options (with -V or --verify):
Install/Upgrade/Erase options:
Common options for all rpm modes and executables:
Options implemented via popt alias/exec:
Help options:
```

Intuitively, it should probably go under "Query options", however, `--query` acts as a _mode selector_ and all the other options in that section are only relevant in that specific mode. So putting it amongst them wouldn't capture this "hierarchy" relationship well. 

Now, one could argue that `--install`, `--erase` etc. are also listed together with their options (under "Install/Upgrade/Erase options"), so we would at least make things more consistent - but IMO, this wouldn't really be an improvement, as it would make the "Query options" section even more confusing, not less.

Another way to solve this, perhaps more consistently, would be to break out `--install`, `--erase`, `--query` (and all the other mode selectors) into a separate help section called something like "Mode options". The problem with that is, we would have to reorganize the popt option tables, the associated `lib/popt*.c` files and the popt callback functions quite a bit. Since this is a critical piece of code (where all the option parsing takes place), I'd rather avoid doing such changes for such little gains.

If we ever want to make the help page better (which I'm all for), I think we should start by reorganizing the option tables and parsing routines first. From that, a good help format would naturally "emerge".

BTW, the above also applies to the `--verify` option which is also hidden.

With that said, I'm closing this issue. If you have an idea how to solve this in a less intrusive way, please reopen or file a new one.

**Comment by xsuchy** [comment]
I have one solution. `poptPrintUsage` is used in two places.  Simply put after each of that use `printModeSelectors` which will be a new function with

```
fprintf(fp, _("[-q] [--verify]\n"));
```

**Comment by dmnks** [comment]
Thanks, Mirku, for looking into this! I like the simplicity of your solution, here's how it would look like:

```
Usage: rpm [-afgplsiv?] [-a|--all] [-f|--file] [-g|--group] [-p|--package] [--pkgid] [--hdrid] [--triggeredby] [--whatconflicts] [--whatrequires] [--whatobsoletes] [--whatprovides]
        [--whatrecommends] [--whatsuggests] [--whatsupplements] [--whatenhances] [--nomanifest] [-c|--configfiles] [-d|--docfiles] [-L|--licensefiles] [-A|--artifactfiles]
        [...]
        [-q|--query] [-v|--verify]
```

That said, I can see two drawbacks:

1) It defeats the purpose of `poptPrintUsage()`
2) Printing the options at the end of the output doesn't help the reading experience

So I'm not sure how I feel about this overall. Tagging @pmatilai to the rescue! :smile:

**Comment by pmatilai** [comment]
Ugh. Just call it a popt limitation and leave it at that.

**Comment by yecril71pl** [comment]
> This is a bit more complicated than one would think. The `--query` option is deliberately hidden from the usage/help output because there simply isn't a good help section to house it:

Usage string does not have any sections.

**Comment by dmnks** [comment]
I didn't mention this explicitly (sorry), but the usage string is generated automatically from the *same data* (popt tables) that the sections in the longer `--help` output are.

**Comment by yecril71pl** [comment]
> Now, one could argue that `--install`, `--erase` etc. are also listed together with their options (under "Install/Upgrade/Erase options"), so we would at least make things more consistent - but IMO, this wouldn't really be an improvement, as it would make the "Query options" section even more confusing, not less.

I consider the following approach less confusing:
````
Query options:
      --query                                            retrieve information about packages
      etc.
````

It would help if we could rearrange the sections so that mode selectors come up first.

**Comment by yecril71pl** [comment]
> Another way to solve this, perhaps more consistently, would be to break out `--install`, `--erase`, `--query` (and all the other mode selectors) into a separate help section called something like "Mode options". The problem with that is, we would have to 

This looks like "the WTF approach" to me.

**Comment by dmnks** [comment]
> I consider the following approach less confusing:
> 
> ```
> Query options:
>       --query                                            retrieve information about packages
>       etc.
> ```
> 
> It would help if we could rearrange the sections so that mode selectors come up first.

You should probably look at it from the perspective of a new user (for whom the help page is primarily meant). What I mean is, if you put `--query` in there, what prevents the user from thinking they could use the other options in that section, separately?

```
Query options (with -q or --query):       
      --query                     retrieve information about packages                                                      
      --dump                         dump basic file information                  
  -l, --list                         list files in package                                 
      --queryformat=QUERYFORMAT      use the following query format              
  -s, --state                        display the states of the listed files               
```

There just isn't the obvious link between `--query` and the other options which only make sense when `--query` is specified. For example, you can't just run `rpm -l` or `rpm -s`.

That being said, the automatic help output has its limitations, some of which are hard limitations of `popt` while others are just the result of the way we structure and feed the data to popt. So by all means, patches are welcome :)

> This looks like "the WTF approach" to me.

Care to elaborate?

**Comment by yecril71pl** [comment]
> There just isn't the obvious link between `--query` and the other options which only make sense when `--query` is specified. For example, you can't just run `rpm -l` or `rpm -s`.

Tru dat.
How about the following approach:
````
  -q, --query                        retrieve information about packages, in particular:
````
> 
> That being said, the automatic help output has its limitations, some of which are hard limitations of `popt` while others are just the result of the way we structure and feed the data to popt. So by all means, patches are welcome :)



> 
> > This looks like "the WTF approach" to me.
> 
> Care to elaborate?

The options are like a hierarchical menu; the menu header should be kept together with the commands in the corresponding submenu, otherwise the users will have to scroll to see the rest of the command and they will get annoyed and tired.

**Comment by dmnks** [comment]
> How about the following approach:
> 
> ```
>   -q, --query                        retrieve information about packages, in particular:

Yes, that would be an improvement.

However, my worry would be that the user would still need to read the description of the first option on the list, in order to comprehend the hierarchy. By quickly scanning the list, it would be quite easy to miss. In this case, there aren't that many query options, so it may not be a problem, but in the case of other sections like "Install/Upgrade/Erase options", this problem would be much more pronounced. (And btw, if we did this, we should also do the same with the "Install/Upgrade/Erase options" section as well, to keep things consistent.)

Another problem is that this relies on an implementation detail where popt just happens to list the options in the same order in which they were defined. Strictly speaking, there's no guarantee that any particular option would be displayed first, which is what this solution relies upon.

Popt just doesn't have support for proper subcommands. You can define a hierarchy of tables, which is the closest you can get to something like subcommands in Python's argparse, however that has its own drawbacks, such as the one you described - that the modes & options would be scattered around and hard to follow.

**Comment by dmnks** [comment]
That said, I don't want to discourage you (or anyone, really) from exploring this in more detail. There certainly are ways to make this better (it is not great at the moment, we all agree).

**Comment by yecril71pl** [comment]
> That said, I don't want to discourage you (or anyone, really) from exploring this in more detail. There certainly are ways to make this better (it is not great at the moment, we all agree).
 
Are you sure this should be Closed?

**Comment by dmnks** [comment]
@pmatilai, do we want to re-open this RFE/bug and keep it open in the long run? The help page has its shortcomings for sure, but I'm not sure if we want to track stuff that's not worked on currently.

**Comment by yecril71pl** [comment]
> @pmatilai, do we want to re-open this RFE/bug and keep it open in the long run? The help page has its shortcomings for sure, but I'm not sure if we want to track stuff that's not worked on currently.

META: **Closed** means to me that it is not going to be fixed even if I submit a patch.  Valid reasons may be: invalid, infeasible, obsolete, unsupported, controversial, by design, upstream etc.

**Comment by dmnks** [comment]
Whether or not a patch will be accepted has nothing to do with the status of this issue.

**Comment by dmnks** [comment]
Anyway, no point in arguing about the proper status. It is a defect which we'd like to fix at some point, so let's reopen. There are higher priority items at the moment, but we do have other such long-standing issues open for tracking purposes, too.

**Comment by pmatilai** [comment]
Like @dmnks said, community is free to look into the matter, whether there are open tickets or not is not relevant. If you care about it deeply then submit a patch, meta-argumentation is waste of everybody's time.

**Comment by yecril71pl** [comment]
> Like @dmnks said, community is free to look into the matter, whether there are open tickets or not is not relevant. If you care about it deeply then submit a patch, meta-argumentation is waste of everybody's time.

Depends on https://github.com/dex4er/fakechroot/pull/93 :-(
